src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/MultiSet.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Collections.Immutable
+{
+    internal class MultiSet<T>
+    {
+        private readonly Dictionary<NullableKeyWrapper, int> _dictionary;
+
+        public MultiSet(IEqualityComparer<T>? equalityComparer)
+        {
+            _dictionary = new Dictionary<NullableKeyWrapper, int>(new NullableKeyWrapperEqualityComparer(equalityComparer ?? EqualityComparer<T>.Default));
+        }
+
+        public void Add(T item)
+        {
+#if NET6_0_OR_GREATER
+            ref int count = ref CollectionsMarshal.GetValueRefOrAddDefault(_dictionary, item, out _);
+            count++;
+#else
+            _dictionary[item] = _dictionary.TryGetValue(item, out int count) ? count + 1 : 1;
+#endif
+        }
+
+        public bool TryRemove(T item)
+        {
+#if NET6_0_OR_GREATER
+            ref int count = ref CollectionsMarshal.GetValueRefOrNullRef(_dictionary, item);
+            if (Unsafe.IsNullRef(ref count) || count == 0)
+#else
+            if (!_dictionary.TryGetValue(item, out int count) || count == 0)
+#endif
+            {
+                return false;
+            }
+
+#if NET6_0_OR_GREATER
+            count--;
+#else
+            _dictionary[item] = count - 1;
+#endif
+            return true;
+        }
+
+        private readonly struct NullableKeyWrapper
+        {
+            public readonly T Key;
+
+            public static implicit operator NullableKeyWrapper(T key)
+            {
+                return new NullableKeyWrapper(key);
+            }
+
+            private NullableKeyWrapper(T key)
+            {
+                Key = key;
+            }
+        }
+
+        private class NullableKeyWrapperEqualityComparer : IEqualityComparer<NullableKeyWrapper>
+        {
+            private readonly IEqualityComparer<T> _keyComparer;
+
+            public NullableKeyWrapperEqualityComparer(IEqualityComparer<T> keyComparer)
+            {
+                _keyComparer = keyComparer;
+            }
+
+            public int GetHashCode(NullableKeyWrapper obj)
+            {
+                return obj.Key == null ? 0 : _keyComparer.GetHashCode(obj.Key);
```

Comment:
The null check would violate custom equality comparers that equate null with non-null keys.

Reply:
Yes, but struggle here is as following, please help decide:

1. See regular [HashSet<T>](https://github.com/dotnet/runtime/blob/c32b1f5c7c476aa7a4391558d88b4adf178f3df7/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/HashSet.cs#L1151), it uses same way to handle `null` (no matter comparer) so it will also hit issue you mentioned unless custom comparer's `GetHashCode(null) == 0` also.
2. See definition of `IEqualityComparer<T>.GetHashCode([DisallowNull] T obj)`: interface declaration not allow null argument and compiler nullable check will also find it; although we know Default compare supports and return 0 at least.

Reply:
> interface declaration not allow null argument and compiler nullable check will also find it; although we know Default compare supports and return 0 at least.

I think that's fine, I would probably just add a suppression. A good guide is the current behavior of `ImmutableArray.Index`, which works with `null` elements if no equality comparer is specified and will surface any null exceptions a custom equality comparer might throw (it only exercises the `Equals` method, not `GetHashCode`).

Reply:
Fixed.